### ISSUE TITLE
WeakValueMap or another algorithm to support Garbage Collection

### ISSUE THREAD

**Issue by btakita** [proposal-signals/proposal-signals#156]
> If possible: A computed Signal should be garbage-collectable if nothing live is referencing it for possible future reads, even if it's linked into a broader graph which stays alive (e.g., by reading a state which remains live). 

I created [rmemo](https://github.com/ctx-core/rmemo). A signal library. One of it's unique features is that it supports automatic Garbage Collection [by using `WeakRef`](https://github.com/ctx-core/ctx-core/blob/main/all/rmemo/index.js).

My frustration with `WeakRef` is that it is fairly slow to create. I would be preferable to have a single global `WeakValueMap` to manage the autosubscriber cache. Instead of instantiating a `WeakRef` for every single signal that has a dependency. An alternative way to manage the state in C/C++ would achieve the same result.

A `WeakValueMap` could be used with a generated reference key object to track the dependent signal stored as the Weak Value.

There is an implementation library called [weak-value-map](https://github.com/HunterKohler/weak-value-map). It instantiates a `WeakRef` for each value, removing the performance benefit.

**Comment by littledan** [comment]
Note that the current signal polyfill does support this weak collection already, through the technique of removing the source->sink backedges for unwatched sink nodes. It ends up not requiring any special GC support at all.

**Comment by btakita** [comment]
Maybe I misunderstood the intent of the quote & the definition of "live":

> If possible: A computed Signal should be garbage-collectable if nothing live is referencing it for possible future reads, even if it's linked into a broader graph which stays alive (e.g., by reading a state which remains live).

Particulary:

> even if it's linked into a broader graph which stays alive

I implemented rmemo to have a Signal be eligible for GC if unreferenced elsewhere. Even if there is a link. i.e. The library doesn't create a strong reference with a link in the graph.

> through the technique of removing the source->sink backedges for unwatched sink nodes.

I read this as: if a sink is connected in a broader graph & not referenced elsewhere, that sink is **not eligable** for GC. Am I correct?

**Comment by alxhub** [comment]
Watchers are by definition owned things - something has to destroy them, otherwise they'll keep notifying. It's similar to registering an event listener on the `window`.

It'd be unusual, but not unheard of, for a watcher to be dropped/leaked without destroying it. In that case, the watcher would be permanent, which _may_ be intended behavior.

**Comment by btakita** [comment]
I have a better grasp on mapping the nomenclatures. In rmemo, a memo (equivalent to a Computed) can also be an effect. So Computed & effect are combined. The memo needs to remain in scope, otherwise it will be Garbage Collected.

So in terms of this Standard Proposal nomenclature. Once an effect leaves scope, it & it's dependencies (not strongly held elsewhere) would instead be subject to GC.

rmemo seems to be unique in this strategy. VanJS has a similar concept predicated on whether the DOM Element is attached to document. One the Element is unattached, it's Signals are subject to removal on the next update & then GC.

I believe that having the effects be subject to GC is more primitive behavior. With most implementations storing a strong reference to the effects. Once an effect is unwatched, it can be deactivated & it's strong reference removed.

**Comment by backbone87** [comment]
you can register the watcher yourself in a finalization registry to auto-unmount. the reference impl in this repo currently wouldn't allow for this since one needs a strong reference to the watcher in order to dispose.

ugly draft:
```
const finalize = new FinalizationRegistry((watcher) => watcher.unwatch(...));
function createAutoDisposedWatcher() {
  const watcher = new Signal.Watcher(...);
  const proxy = new Proxy(watcher, Reflect);
  finalize.register(proxy, watcher)

  return proxy
}
```

**Comment by alxhub** [comment]
Having watchers that automatically shut down when they go out of scope feels very unpredictable and error-prone to me. Consider all of the warnings on [`FinalizationRegistry`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/FinalizationRegistry#avoid_where_possible) about the lack of guarantees and potential differences across engines, etc. In particular:

> Developers shouldn't rely on cleanup callbacks for essential program logic.

And:

> A conforming JavaScript implementation, even one that does garbage collection, is not required to call cleanup callbacks. When and whether it does so is entirely down to the implementation of the JavaScript engine. When a registered object is reclaimed, any cleanup callbacks for it may be called then, or some time later, or not at all.

**Comment by btakita** [comment]
> you can register the watcher yourself in a finalization registry to auto-unmount. the reference impl in this repo currently wouldn't allow for this since one needs a strong reference to the watcher in order to dispose.

rmemo is a tiny fish here in a large lake. In my dream world, the watch function could take a `WeakRef` or better yet an option to store on a `WeakValueMap` (which would be require an addition to the standards).

> Having watchers that automatically shut down when they go out of scope feels very unpredictable and error-prone to me.

I've been using this technique since October 2023. I have mostly experienced simplicity. I had one GC issue when implementing a helper function to chain cancelable promises that wait for a Signal's condition to be met. But otherwise, it's been a relatively simple matter to attach the Signal to an object with the correct scope.

> A conforming JavaScript implementation, even one that does garbage collection, is not required to call cleanup callbacks. When and whether it does so is entirely down to the implementation of the JavaScript engine. When a registered object is reclaimed, any cleanup callbacks for it may be called then, or some time later, or not at all.

If a Signal is lazy, what is the difference? The watcher loop would longer than optimal if GC never happens. In practice, I've used the major browser engines, nodejs, & bun. GC seems to happen within a reasonable amount of time. Granted, someone could turn off GC. In that case, there is explicit unwatch.

**Comment by btakita** [comment]
@littledan @alxhub 

JS is a Garbage Collected language. With the UI library, an unload event can be listened to unwatch the signal. Outside of a UI library, there is no such event. WeakRef is the "least bad" option that I can think of. Perhaps someone else has a better alternative? I'm all ears.

**Comment by shaylew** [comment]
This seems to be two similar but separate issues:
1. Should a watcher strongly retain its sources? 
2. Should a watcher that isn't referenced, except through the graph, be disposed of and stop watching/firing its callback?

(2) is the one that exposes finalization timing behavior. If the watched signals are still alive and still changing, when exactly the watcher stops running would be up to the gc. This is unpredictable, and probably undesirable; all corners of the JS spec that have this nondeterminism baked in carry a serious warning label and are (rightfully) under heavy scrutiny.

(1) has no such complication, because a source whose only remaining reference is from a watcher can't change. So I think we should probably arrange for watchers to not retain their unchanged dependences. _Changed_ dependencies may be necessary to keep track of, in order to re-watch them; so I imagine the implementation here would look like maintaining only the `notifiedBy` list on the watcher, rather than the full set of sources.

**Comment by alxhub** [comment]
> Should a watcher strongly retain its sources?

At least with "correct" usage, I think watchers will never have the only remaining reference to a source, because in order to be a source in the first place, it'd have to be reachable within the computation being watched (part of its closure).

The only way for this _not_ to be true is if the computation function changes in a way that:
* affects the result of the computation (a source which was read during the computation would no longer be read)
* does not change any _other_ dependencies of the computation in a way where the watcher would schedule a rerun\

**Comment by shaylew** [comment]
@alxhub That's all true of Compteds, but Watchers in this proposal don't have a tracked body like that -- they can add/remove edges with `watch` and `unwatch` however they like. 

Definitely more of an edge case in some usage patterns than others, but I think it might come up in practice decently often. Some frameworks are likely to use a pattern where the effect body is a Computed and downstream of it is a Watcher used for scheduling, and in that setup there may be no other reference to the effect (_or_ its closure body) except through the watcher.

**Comment by alxhub** [comment]
@shaylew ahhh, yeah, I was talking about a dependency that was upstream of a watched `Computed`. I see what you mean - if you watch a `Signal.State` and every other reference gets dropped, it can no longer change.

**Comment by btakita** [comment]
Hi all, I wrote a blog post titled [Signals Proposal Compared to rmemo/WeakRef](https://briantakita.me/posts/signals-proposal-compared-to-rmemo-weak-ref). The blog post follows up @littledan's performance concern brought up on Discord. And goes more in depth on the rationale on how WeakRef simplifies & makes domain apis more modular. There is a [discussion thread](https://discord.com/channels/1224208557126193333/1236762609676452001) on the Discord Chat as well.
